src/posterior_gp/approx_posterior_gp.jl

```diff
@@ -107,54 +151,140 @@ function update_approx_posterior(f_post_approx::ApproxPosteriorGP, u::FiniteGP)
-    A = f.data.U' \ cov(f.prior, f.data.z, x)
+function StatsBase.mean_and_cov(f::ApproxPosteriorGP{<:VFE}, x::AbstractVector)
+    A = f.data.U' \ cov(f.prior, inducing_points(f), x)
     m_post = mean(f.prior, x) + A' * f.data.m_ε
     C_post = cov(f.prior, x) - At_A(A) + Xt_invA_X(f.data.Λ_ε, A)
     return m_post, C_post
 end
 
-function StatsBase.mean_and_var(f::ApproxPosteriorGP{VFE}, x::AbstractVector)
-    A = f.data.U' \ cov(f.prior, f.data.z, x)
+function StatsBase.mean_and_var(f::ApproxPosteriorGP{<:VFE}, x::AbstractVector)
+    A = f.data.U' \ cov(f.prior, inducing_points(f), x)
     m_post = mean(f.prior, x) + A' * f.data.m_ε
     c_post = var(f.prior, x) - diag_At_A(A) + diag_Xt_invA_X(f.data.Λ_ε, A)
     return m_post, c_post
 end
+
+inducing_points(f::ApproxPosteriorGP{<:VFE}) = f.approx.z
+
+"""
+    elbo(v::VFE, fx::FiniteGP, y::AbstractVector{<:Real})
+
+The Titsias Evidence Lower BOund (ELBO) [1]. `y` are observations of `fx`, and `v.z`
+are inducing points.
+
+
+```jldoctest
+julia> f = GP(Matern52Kernel());
+
+julia> x = randn(1000);
+
+julia> z = range(-5.0, 5.0; length=13);
+
+julia> v = VFE(z);
+
+julia> y = rand(f(x, 0.1));
+
+julia> elbo(v, f(x, 0.1), y) < logpdf(f(x, 0.1), y)
+true
+```
+
+[1] - M. K. Titsias. "Variational learning of inducing variables in sparse Gaussian
+processes". In: Proceedings of the Twelfth International Conference on Artificial
+Intelligence and Statistics. 2009.
+"""
+function elbo(v::VFE, fx::FiniteGP, y::AbstractVector{<:Real})
+    _dtc, A = _compute_intermediates(fx, y, fx.f(v.z, v.jitter))
+    return _dtc - (tr_Cf_invΣy(fx, fx.Σy) - sum(abs2, A)) / 2
+end
+
+"""
+    dtc(v::VFE, fx::FiniteGP, y::AbstractVector{<:Real})
+
+The Deterministic Training Conditional (DTC) [1]. `y` are observations of `fx`, and `v.z`
+are inducing points.
+
+
+```jldoctest
+julia> f = GP(Matern52Kernel());
+
+julia> x = randn(1000);
+
+julia> z = range(-5.0, 5.0; length=256);
+
+julia> v = VFE(z);
+
+julia> y = rand(f(x, 0.1));
+
+julia> isapprox(dtc(v, f(x, 0.1), y), logpdf(f(x, 0.1), y); atol=1e-6, rtol=1e-6)
+true
+```
+
+[1] - M. Seeger, C. K. I. Williams and N. D. Lawrence. "Fast Forward Selection to Speed Up
+Sparse Gaussian Process Regression". In: Proceedings of the Ninth International Workshop on
+Artificial Intelligence and Statistics. 2003
+"""
+function dtc(v::VFE, fx::FiniteGP, y::AbstractVector{<:Real})
+    return first(_compute_intermediates(fx, y, fx.f(v.z, v.jitter)))
+end
+
+# Factor out computations common to the `elbo` and `dtc`.
+function _compute_intermediates(fx::FiniteGP, y::AbstractVector{<:Real}, fz::FiniteGP)
+    consistency_check(fx, y)
+    chol_Σy = _cholesky(fx.Σy)
+
+    A = cholesky(_symmetric(cov(fz))).U' \ (chol_Σy.U' \ cov(fx, fz))'
+    Λ_ε = cholesky(Symmetric(A * A' + I))
+    δ = chol_Σy.U' \ (y - mean(fx))
+
+    tmp = logdet(chol_Σy) + logdet(Λ_ε) + sum(abs2, δ) - sum(abs2, Λ_ε.U' \ (A * δ))
+    _dtc = -(length(y) * typeof(tmp)(log(2π)) + tmp) / 2
```

Comment:
Is the cast to `typeof(tmp)` necessary?
Also, is there a reason to use `log(2π)` here instead of the `log2π` constant? (that was already providing us a bit of grief with StatsFuns...)

Reply:
Ah I might've confused this with the KernelFunctions repo. But the question still stands (probably at @willtebbutt) and depending on the answer, maybe we should depend on IrrationalConstants here as well then?

Reply:
Happy for this to change to use IrrationalConstants -- the reason to cast it was to ensure that e.g. `Float32`s don't accidentally get promoted to `Float64`s, so as long as we maintain that functionality, I don't particularly mind. (I'm not sure we actually test that that works though...)

Reply:
Yeah that makes sense that you don't want to accidentally promote due to internal constants. I think
```suggestion
    _dtc = -(length(y) * log2π) + tmp) / 2
```
would be rather more pleasant to read, and KernelFunctions already depends on IrrationalConstants so it's not adding overhead:)

Reply:
changed it to use IrrationalConstants - I think you still need the cast though because `length(y) * log2π isa Float64` unless there's some way around that?